File: packages/java/anip-graphql/src/main/java/dev/anip/graphql/GraphQLResponseMapper.java
package dev.anip.graphql;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Maps ANIP snake_case invoke results to GraphQL camelCase response shape.
 */
public class GraphQLResponseMapper {

    private GraphQLResponseMapper() {}

    /**
     * Converts snake_case to camelCase.
     * e.g. "cost_actual" -> "costActual"
     */
    public static String toCamelCase(String snake) {
        if (snake == null || snake.isEmpty()) return snake;
        String[] parts = snake.split("_");
        StringBuilder sb = new StringBuilder(parts[0]);
        for (int i = 1; i < parts.length; i++) {
            if (!parts[i].isEmpty()) {
                sb.append(Character.toUpperCase(parts[i].charAt(0)));
                sb.append(parts[i].substring(1));
            }
        }
        return sb.toString();
    }

    /**
     * Converts camelCase to snake_case.
     * e.g. "searchFlights" -> "search_flights"
     */
    public static String toSnakeCase(String camel) {
        if (camel == null || camel.isEmpty()) return camel;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < camel.length(); i++) {
            char c = camel.charAt(i);
            if (Character.isUpperCase(c)) {
                if (i > 0) sb.append('_');
                sb.append(Character.toLowerCase(c));
            } else {
                sb.append(c);
            }
        }
        return sb.toString();
    }

    /**
     * Converts snake_case to PascalCase.
     */
    public static String toPascalCase(String snake) {
        if (snake == null || snake.isEmpty()) return snake;
        String[] parts = snake.split("_");
        StringBuilder sb = new StringBuilder();
        for (String part : parts) {
            if (!part.isEmpty()) {
                sb.append(Character.toUpperCase(part.charAt(0)));
                sb.append(part.substring(1));
            }
        }
        return sb.toString();
    }

    /**
     * Maps an ANIP invoke result (snake_case) to GraphQL result shape (camelCase).
     */
    @SuppressWarnings("unchecked")
    public static Map<String, Object> buildGraphQLResponse(Map<String, Object> result) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", result.getOrDefault("success", false));
        response.put("result", result.get("result"));
        response.put("costActual", null);
        response.put("failure", null);

        Object costActualRaw = result.get("cost_actual");
        if (costActualRaw instanceof Map) {
            Map<String, Object> costMap = (Map<String, Object>) costActualRaw;
            Map<String, Object> mapped = new LinkedHashMap<>();
            mapped.put("financial", costMap.get("financial"));
            mapped.put("varianceFromEstimate", costMap.get("variance_from_estimate"));
            response.put("costActual", mapped);
        }

        Object failureRaw = result.get("failure");
        if (failureRaw instanceof Map) {
            Map<String, Object> failure = (Map<String, Object>) failureRaw;
            Map<String, Object> f = new LinkedHashMap<>();
            f.put("type", failure.getOrDefault("type", "unknown"));
            f.put("detail", failure.getOrDefault("detail", ""));
            f.put("resolution", null);
            f.put("retry", failure.getOrDefault("retry", false));

            Object resolutionRaw = failure.get("resolution");
            if (resolutionRaw instanceof Map) {
                Map<String, Object> res = (Map<String, Object>) resolutionRaw;
                Map<String, Object> mappedRes = new LinkedHashMap<>();
                mappedRes.put("action", res.get("action"));
                mappedRes.put("requires", res.get("requires"));
                mappedRes.put("grantableBy", res.get("grantable_by"));
                f.put("resolution", mappedRes);
            }

            response.put("failure", f);
        }

        return response;
    }
}

File: packages/csharp/src/Anip.GraphQL/GraphQLResponseMapper.cs
namespace Anip.GraphQL;

/// <summary>
/// Maps ANIP snake_case invoke results to GraphQL camelCase response shape.
/// </summary>
public static class GraphQLResponseMapper
{
    /// <summary>
    /// Converts snake_case to camelCase.
    /// e.g. "cost_actual" -> "costActual"
    /// </summary>
    public static string ToCamelCase(string snake)
    {
        if (string.IsNullOrEmpty(snake)) return snake;
        var parts = snake.Split('_');
        var sb = new System.Text.StringBuilder(parts[0]);
        for (int i = 1; i < parts.Length; i++)
        {
            if (parts[i].Length > 0)
            {
                sb.Append(char.ToUpper(parts[i][0]));
                sb.Append(parts[i][1..]);
            }
        }
        return sb.ToString();
    }

    /// <summary>
    /// Converts camelCase to snake_case.
    /// e.g. "searchFlights" -> "search_flights"
    /// </summary>
    public static string ToSnakeCase(string camel)
    {
        if (string.IsNullOrEmpty(camel)) return camel;
        var sb = new System.Text.StringBuilder();
        for (int i = 0; i < camel.Length; i++)
        {
            var c = camel[i];
            if (char.IsUpper(c))
            {
                if (i > 0) sb.Append('_');
                sb.Append(char.ToLower(c));
            }
            else
            {
                sb.Append(c);
            }
        }
        return sb.ToString();
    }

    /// <summary>
    /// Converts snake_case to PascalCase.
    /// </summary>
    public static string ToPascalCase(string snake)
    {
        if (string.IsNullOrEmpty(snake)) return snake;
        var parts = snake.Split('_');
        var sb = new System.Text.StringBuilder();
        foreach (var part in parts)
        {
            if (part.Length > 0)
            {
                sb.Append(char.ToUpper(part[0]));
                sb.Append(part[1..]);
            }
        }
        return sb.ToString();
    }

    /// <summary>
    /// Maps an ANIP invoke result (snake_case) to GraphQL result shape (camelCase).
    /// </summary>
    public static Dictionary<string, object?> BuildGraphQLResponse(Dictionary<string, object?> result)
    {
        var response = new Dictionary<string, object?>
        {
            ["success"] = result.GetValueOrDefault("success", false),
            ["result"] = result.GetValueOrDefault("result"),
            ["costActual"] = null,
            ["failure"] = null,
        };

        if (result.TryGetValue("cost_actual", out var costActualRaw) && costActualRaw is Dictionary<string, object?> costMap)
        {
            response["costActual"] = new Dictionary<string, object?>
            {
                ["financial"] = costMap.GetValueOrDefault("financial"),
                ["varianceFromEstimate"] = costMap.GetValueOrDefault("variance_from_estimate"),
            };
        }

        if (result.TryGetValue("failure", out var failureRaw) && failureRaw is Dictionary<string, object?> failure)
        {
            var f = new Dictionary<string, object?>
            {
                ["type"] = failure.GetValueOrDefault("type", "unknown"),
                ["detail"] = failure.GetValueOrDefault("detail", ""),
                ["resolution"] = null,
                ["retry"] = failure.GetValueOrDefault("retry", false),
            };

            if (failure.TryGetValue("resolution", out var resolutionRaw) && resolutionRaw is Dictionary<string, object?> res)
            {
                f["resolution"] = new Dictionary<string, object?>
                {
                    ["action"] = res.GetValueOrDefault("action"),
                    ["requires"] = res.GetValueOrDefault("requires"),
                    ["grantableBy"] = res.GetValueOrDefault("grantable_by"),
                };
            }

            response["failure"] = f;
        }

        return response;
    }
}
